In [1]:
%load_ext autoreload
%autoreload 2

In [2]:
import pandas as pd

# Load metadata.csv
metadata_df = pd.read_csv("../data/raw/metadata.csv")
metadata_df

feature       name                                        description
0         1       team                                       Name of team
1         2       conf                                 Name of conference
2         3         GP                                       Games played
3         4    Min_per  Player's percentage of available team minutes ...
4         5       ORtg  ORtg - Offensive Rating (available since the 1...
..      ...        ...                                                ...
59       66        stl  STL - Steals (available since the 1973-74 seas...
60       67        blk  BLK - Blocks (available since the 1973-74 seas...
61       68        pts                                       PTS - Points
62       69  player_id                        Unique identifier of player
63       70    drafted  Target - Was the player drafted at the end of ...

[64 rows x 3 columns]

In [3]:
# Load train.csv
train_df = pd.read_csv("../data/raw/train.csv")

# Display the first few rows of the train.csv data
train_df.head()

C:\Users\Sahil Kotak\AppData\Local\Temp\ipykernel_1592\3403509580.py:2: DtypeWarning: Columns (26) have mixed types. Specify dtype option on import or set low_memory=False.
  train_df = pd.read_csv("../data/raw/train.csv")


team conf  GP  Min_per   Ortg   usg   eFG  TS_per  ORB_per   
0  South Alabama   SB  26     29.5   97.3  16.6  42.5   44.43      1.6  \
1       Utah St.  WAC  34     60.9  108.3  14.9  52.4   54.48      3.8   
2  South Florida   BE  27     72.0   96.2  21.8  45.7   47.98      2.1   
3     Pepperdine  WCC  30     44.5   97.7  16.0  53.6   53.69      4.1   
4        Pacific   BW  33     56.2   96.5  22.0  52.8   54.31      8.3   

   DRB_per  ...     dgbpm    oreb    dreb    treb     ast     stl     blk   
0      4.6  ... -1.941150  0.1923  0.6154  0.8077  1.1923  0.3462  0.0385  \
1      6.3  ... -0.247934  0.6765  1.2647  1.9412  1.8235  0.4118  0.2353   
2      8.0  ... -0.883163  0.6296  2.3333  2.9630  1.9630  0.4815  0.0000   
3      9.4  ... -0.393459  0.7000  1.4333  2.1333  1.1000  0.5667  0.1333   
4     18.6  ... -0.668318  1.4242  3.3030  4.7273  0.8485  0.4545  0.3333   

       pts                             player_id  drafted  
0   3.8846  7be2aead-da4e-4d13-a74b-4c1e692e2368      0.0  
1   5.9412  61de55d9-1582-4ea4-b593-44f6aa6524a6      0.0  
2  12.1852  efdc4cfc-9dd0-4bf8-acef-7273e4d5b655      0.0  
3   4.9333  14f05660-bb3c-4868-b3dd-09bcdb64279d      0.0  
4   7.5758  a58db52f-fbba-4e7b-83d0-371efcfed039      0.0  

[5 rows x 64 columns]

In [4]:
train_df.shape

(56091, 64)

In [5]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56091 entries, 0 to 56090
Data columns (total 64 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   team                 56091 non-null  object 
 1   conf                 56091 non-null  object 
 2   GP                   56091 non-null  int64  
 3   Min_per              56091 non-null  float64
 4   Ortg                 56091 non-null  float64
 5   usg                  56091 non-null  float64
 6   eFG                  56091 non-null  float64
 7   TS_per               56091 non-null  float64
 8   ORB_per              56091 non-null  float64
 9   DRB_per              56091 non-null  float64
 10  AST_per              56091 non-null  float64
 11  TO_per               56091 non-null  float64
 12  FTM                  56091 non-null  int64  
 13  FTA                  56091 non-null  int64  
 14  FT_per               56091 non-null  float64
 15  twoPM                56091 non-null 

In [6]:
train_df.describe(include='all')

team   conf            GP       Min_per          Ortg           usg   
count   56091  56091  56091.000000  56091.000000  56091.000000  56091.000000  \
unique    358     36           NaN           NaN           NaN           NaN   
top      Army    ACC           NaN           NaN           NaN           NaN   
freq      209   2297           NaN           NaN           NaN           NaN   
mean      NaN    NaN     23.411973     37.325229     91.858295     18.149265   
std       NaN    NaN     10.090542     28.061897     30.538819      6.204489   
min       NaN    NaN      1.000000      0.000000      0.000000      0.000000   
25%       NaN    NaN     16.000000      9.500000     84.100000     14.500000   
50%       NaN    NaN     28.000000     35.900000     97.100000     18.100000   
75%       NaN    NaN     31.000000     62.200000    106.900000     21.800000   
max       NaN    NaN     41.000000     98.000000    834.000000     50.000000   

                 eFG        TS_per       ORB_per       DRB_per  ...   
count   56091.000000  56091.000000  56091.000000  56091.000000  ...  \
unique           NaN           NaN           NaN           NaN  ...   
top              NaN           NaN           NaN           NaN  ...   
freq             NaN           NaN           NaN           NaN  ...   
mean       44.575513     47.676882      5.575376     12.673716  ...   
std        18.213813     17.414133      9.480239     10.878099  ...   
min         0.000000      0.000000      0.000000      0.000000  ...   
25%        40.100000     43.930000      1.800000      8.400000  ...   
50%        47.700000     50.850000      4.300000     11.900000  ...   
75%        53.200000     56.060000      8.100000     16.000000  ...   
max       150.000000    150.000000   1576.600000   1385.000000  ...   

               dgbpm          oreb          dreb          treb           ast   
count   56047.000000  56053.000000  56053.000000  56053.000000  56053.000000  \
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean       -0.442297      0.770611      1.892593      2.663204      1.073038   
std         3.279438      0.738899      1.479148      2.099913      1.171148   
min      -100.984000      0.000000      0.000000      0.000000      0.000000   
25%        -1.728030      0.225800      0.700000      1.000000      0.222200   
50%        -0.313051      0.545500      1.633300      2.259300      0.677400   
75%         1.066415      1.111100      2.758600      3.833300      1.533300   
max        78.985500      5.933300     11.545500     14.533300     10.000000   

                 stl           blk           pts   
count   56053.000000  56053.000000  56053.000000  \
unique           NaN           NaN           NaN   
top              NaN           NaN           NaN   
freq             NaN           NaN           NaN   
mean        0.529475      0.282825      5.775381   
std         0.469152      0.417754      4.953723   
min         0.000000      0.000000      0.000000   
25%         0.160000      0.000000      1.571400   
50%         0.424200      0.133300      4.482800   
75%         0.793100      0.366700      9.069000   
max         4.000000      5.258100     30.090900   

                                   player_id       drafted  
count                                  56091  56091.000000  
unique                                 23929           NaN  
top     06dc8dc2-888e-4941-9106-4798cddfb9d2           NaN  
freq                                       6           NaN  
mean                                     NaN      0.009556  
std                                      NaN      0.097287  
min                                      NaN      0.000000  
25%                                      NaN      0.000000  
50%                                      NaN      0.000000  

## Data Cleaning

In [7]:
# Check for missing values in the dataset
missing_values = train_df.isnull().sum()

# Display columns with missing values and their counts
missing_values[missing_values > 0]

yr                       292
ht                        98
num                     4690
Rec_Rank               39055
ast_tov                 4190
rimmade                 6081
rimmade_rimmiss         6081
midmade                 6081
midmade_midmiss         6081
rim_ratio               9464
mid_ratio               9688
dunksmade               6081
dunksmiss_dunksmade     6081
dunks_ratio            30793
pick                   54705
drtg                      44
adrtg                     44
dporpag                   44
stops                     44
bpm                       44
obpm                      44
dbpm                      44
gbpm                      44
mp                        38
ogbpm                     44
dgbpm                     44
oreb                      38
dreb                      38
treb                      38
ast                       38
stl                       38
blk                       38
pts                       38
dtype: int64

In [8]:
# Calculate the percentage of missing values for each column
missing_percentage = (missing_values / len(train_df)) * 100

# Display columns with missing value percentages
missing_percentage[missing_percentage > 0].sort_values(ascending=False)

pick                   97.529015
Rec_Rank               69.627926
dunks_ratio            54.898290
mid_ratio              17.271933
rim_ratio              16.872582
rimmade                10.841311
rimmade_rimmiss        10.841311
midmade                10.841311
midmade_midmiss        10.841311
dunksmade              10.841311
dunksmiss_dunksmade    10.841311
num                     8.361413
ast_tov                 7.470004
yr                      0.520583
ht                      0.174716
obpm                    0.078444
dgbpm                   0.078444
ogbpm                   0.078444
gbpm                    0.078444
dbpm                    0.078444
adrtg                   0.078444
bpm                     0.078444
stops                   0.078444
dporpag                 0.078444
drtg                    0.078444
mp                      0.067747
oreb                    0.067747
dreb                    0.067747
treb                    0.067747
ast                     0.067747
stl       

Based on the percentage of missing values, we can make the following decisions:

Drop Columns: We can drop columns pick and Rec_Rank as they have a high percentage of missing values (97.53% and 69.63% respectively). The dunks_ratio column also has over 50% missing values, so we can consider dropping it as well.

Impute Values: For columns with a lower percentage of missing values, we can impute these values. For numerical columns, we'll use the median, and for categorical columns (like yr and ht), we'll use the mode.

Drop Rows: Since the percentage of missing values for some columns is very low (less than 0.1%), we can drop those rows.

In [9]:
# 1. Drop columns with a high percentage of missing values
columns_to_drop = ['pick', 'Rec_Rank', 'dunks_ratio']
train_df = train_df.drop(columns=columns_to_drop)

In [12]:
# 2. Impute missing values
# For numerical columns
numerical_columns_to_impute = ['ast_tov', 'rimmade', 'rimmade_rimmiss', 'midmade', 'midmade_midmiss', 
                               'dunksmade', 'dunksmiss_dunksmade', 'num', 'drtg', 'adrtg', 'dporpag', 'stops', 
                               'bpm', 'obpm', 'dbpm', 'gbpm', 'mp', 'ogbpm', 'dgbpm', 'oreb', 'dreb', 'treb', 
                               'ast', 'stl', 'blk', 'pts']
#for column in numerical_columns_to_impute:
#    train_df[column] = train_df[column].fillna(train_df[column].median())
#
# For categorical columns
#categorical_columns_to_impute = ['yr', 'ht']
#for column in categorical_columns_to_impute:
#    train_df[column] = train_df[column].fillna(train_df[column].mode()[0])

In [13]:
# Check unique values in the columns we assumed to be numerical
non_numeric_values = {}

for column in numerical_columns_to_impute:
    try:
        # Try converting the column to float
        train_df[column].astype(float)
    except ValueError:
        # If there's an error, store the unique values of the column
        non_numeric_values[column] = train_df[column].unique()

non_numeric_values

{'num': array([nan, '53', '5', '14', '22', '20', '34', '24', '54', '15', '13',
        '10', '21', '32', '12', '51', '31', '33', '1', '45', '50', '3',
        '23', '2', '4', '55', '35', '42', '25', '41', '52', '11', '40',
        '30', '43', '0', '44', '23B', 2.0, 15.0, 11.0, 0.0, 5.0, 34.0,
        24.0, 22.0, 1.0, 32.0, 23.0, 14.0, 20.0, 4.0, 33.0, 25.0, 40.0,
        43.0, 3.0, 12.0, 10.0, 30.0, 21.0, 31.0, 13.0, 50.0, 44.0, 41.0,
        35.0, 42.0, 45.0, 55.0, 54.0, 53.0, 52.0, 51.0, 8.0, 26.0, '99',
        '4A', '31/24', '--', 99.0], dtype=object)}

The num column, which we assumed to be numerical, contains non-numeric values such as '23B', 'None', '4A', '31/24', and '--'.

1. Convert any value that can be cast to a number.
2. Impute the non-numeric values and NaNs with the median of the numeric values.
3. Proceed with imputation for the other columns.

In [14]:
# Convert values in 'num' column to numeric where possible, setting errors='coerce' to replace non-numeric values with NaNs
train_df['num'] = pd.to_numeric(train_df['num'], errors='coerce')

# Impute NaNs in 'num' column with its median
train_df['num'] = train_df['num'].fillna(train_df['num'].median())

# Now, proceed with imputation for the other columns
for column in numerical_columns_to_impute:
    train_df[column] = train_df[column].fillna(train_df[column].median())

# Check again for missing values
missing_values_after_cleaning = train_df.isnull().sum()
missing_values_after_cleaning[missing_values_after_cleaning > 0]

yr            292
ht             98
rim_ratio    9464
mid_ratio    9688
dtype: int64

We've handled the non-numeric values in the num column. However, there are still some missing values in the columns yr, ht, rim_ratio, and mid_ratio.

We will impute yr and ht with their modes. For rim_ratio and mid_ratio, which are numerical columns, we'll use their medians for imputation.

In [15]:
# Impute 'yr' and 'ht' with their modes
train_df['yr'] = train_df['yr'].fillna(train_df['yr'].mode()[0])
train_df['ht'] = train_df['ht'].fillna(train_df['ht'].mode()[0])

# Impute 'rim_ratio' and 'mid_ratio' with their medians
train_df['rim_ratio'] = train_df['rim_ratio'].fillna(train_df['rim_ratio'].median())
train_df['mid_ratio'] = train_df['mid_ratio'].fillna(train_df['mid_ratio'].median())

# Check again for missing values
final_missing_values = train_df.isnull().sum()
final_missing_values[final_missing_values > 0]

Series([], dtype: int64)

## Feature Engineering

In [16]:
from sklearn.model_selection import train_test_split

# Split the data into training and validation sets (80% training, 20% validation)
train_set, validation_set = train_test_split(train_df, test_size=0.2, random_state=42, stratify=train_df['drafted'])

train_set.shape, validation_set.shape

((44872, 61), (11219, 61))

We will proceed with the Feature Engineering phase:

1. Encoding Categorical Variables: We'll identify categorical columns and use one-hot encoding or label encoding to convert them into numerical values.
2. Normalizing Numerical Features: This will help ensure that all features are on the same scale, which can be crucial for certain algorithms.

In [17]:
# Identify categorical columns
categorical_columns = train_set.select_dtypes(include=['object']).columns.tolist()

# Exclude 'player_id' from the categorical columns as it's an identifier
categorical_columns.remove('player_id')

# One-hot encode the categorical columns
train_encoded = pd.get_dummies(train_set, columns=categorical_columns, drop_first=True)
validation_encoded = pd.get_dummies(validation_set, columns=categorical_columns, drop_first=True)

# Ensure that both training and validation datasets have the same columns after encoding
missing_cols = set(train_encoded.columns) - set(validation_encoded.columns)
for col in missing_cols:
    validation_encoded[col] = 0

# Align the order of columns in validation_encoded to match train_encoded
validation_encoded = validation_encoded[train_encoded.columns]

train_encoded.shape, validation_encoded.shape

((44872, 490), (11219, 490))

Next we will normalize the numerical features. For this step, we'll use a MinMaxScaler to scale each feature to the range [0, 1]. 

We'll fit the scaler on the training data and then apply it to both the training and validation datasets.

In [18]:
from sklearn.preprocessing import MinMaxScaler

# Exclude target and identifier columns from scaling
columns_to_exclude = ['player_id', 'drafted']
columns_to_scale = [col for col in train_encoded.columns if col not in columns_to_exclude]

# Initialize and fit the scaler on the training data
scaler = MinMaxScaler()
scaler.fit(train_encoded[columns_to_scale])

# Apply the scaler to the training and validation datasets
train_encoded[columns_to_scale] = scaler.transform(train_encoded[columns_to_scale])
validation_encoded[columns_to_scale] = scaler.transform(validation_encoded[columns_to_scale])

# Check the first few rows of the normalized training dataset
train_encoded.head()


GP   Min_per      Ortg    usg       eFG    TS_per   ORB_per   
14530  0.750  0.606122  0.123741  0.428  0.312000  0.338533  0.003633  \
17837  0.600  0.333673  0.123501  0.356  0.355333  0.357267  0.011327   
3172   0.725  0.586735  0.119784  0.294  0.298667  0.351733  0.015815   
18612  0.725  0.592857  0.118705  0.266  0.342000  0.351000  0.018166   
20107  0.425  0.107143  0.090647  0.208  0.311333  0.311133  0.004061   

        DRB_per  AST_per  TO_per  ...  ht_8-Jun  ht_8-May  ht_9-Jun  ht_9-May   
14530  0.014916    0.262   0.146  ...       0.0       0.0       0.0       0.0  \
17837  0.022068    0.101   0.178  ...       0.0       0.0       0.0       0.0   
3172   0.032693    0.050   0.216  ...       0.0       0.0       1.0       0.0   
18612  0.032693    0.034   0.211  ...       0.0       0.0       0.0       0.0   
20107  0.005108    0.138   0.434  ...       0.0       0.0       0.0       0.0   

       ht_Apr-00  ht_Fr  ht_Jr  ht_Jul-00  ht_Jun-00  ht_So  
14530        0.0    0.0    0.0        0.0        0.0    0.0  
17837        0.0    0.0    0.0        0.0        0.0    0.0  
3172         0.0    0.0    0.0        0.0        0.0    0.0  
18612        0.0    0.0    0.0        0.0        0.0    0.0  
20107        0.0    0.0    0.0        0.0        1.0    0.0  

[5 rows x 490 columns]

## Modelling

In [19]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score

# Prepare features and target variable for training and validation datasets
X_train = train_encoded.drop(columns=['player_id', 'drafted'])
y_train = train_encoded['drafted']
X_validation = validation_encoded.drop(columns=['player_id', 'drafted'])
y_validation = validation_encoded['drafted']

# Initialize and train the logistic regression model
logreg = LogisticRegression(max_iter=1000, random_state=42)
logreg.fit(X_train, y_train)

# Predict probabilities on the validation dataset
y_pred_prob = logreg.predict_proba(X_validation)[:, 1]

# Calculate the AUROC score
auroc_score = roc_auc_score(y_validation, y_pred_prob)
auroc_score

0.9895869078137302

# Generating Predictions

In [32]:
test_df = pd.read_csv("../data/raw/test.csv")

# Drop columns with a high percentage of missing values
test_df = test_df.drop(columns=columns_to_drop)

# Impute missing values
# For numerical columns
for column in numerical_columns_to_impute:
    test_df[column] = test_df[column].fillna(test_df[column].median())

# For categorical columns
#for column in categorical_columns_to_impute:
#    test_df[column] = test_df[column].fillna(test_df[column].mode()[0])

In [33]:
# Convert values in 'num' column to numeric where possible, setting errors='coerce' to replace non-numeric values with NaNs
test_df['num'] = pd.to_numeric(test_df['num'], errors='coerce')

# Impute NaNs in 'num' column with its median
test_df['num'] = test_df['num'].fillna(test_df['num'].median())

# Impute 'rim_ratio' and 'mid_ratio' with their medians
test_df['rim_ratio'] = test_df['rim_ratio'].fillna(train_df['rim_ratio'].median())
test_df['mid_ratio'] = test_df['mid_ratio'].fillna(train_df['mid_ratio'].median())

# Now, proceed with imputation for the other columns
for column in numerical_columns_to_impute:
    test_df[column] = test_df[column].fillna(test_df[column].median())

In [34]:
# One-hot encode the categorical columns
test_encoded = pd.get_dummies(test_df, columns=categorical_columns, drop_first=True)

In [35]:
# Ensure that both training and test datasets have the same columns after encoding
missing_cols = set(train_encoded.columns) - set(test_encoded.columns)
for col in missing_cols:
    test_encoded[col] = 0

In [36]:
# Align the order of columns in test_encoded to match train_encoded
test_encoded = test_encoded[train_encoded.columns]

In [37]:
# Apply the scaler to the test dataset
test_encoded[columns_to_scale] = scaler.transform(test_encoded[columns_to_scale])

# Prepare features for test dataset
X_test = test_encoded.drop(columns=['player_id', 'drafted'])

# Generate predictions on the test dataset
y_test_pred_prob = logreg.predict_proba(X_test)[:, 1]

In [38]:
# Create a DataFrame with player_id and predictions
predictions_df = pd.DataFrame({'player_id': test_encoded['player_id'], 'drafted': y_test_pred_prob})

# Generate a csv file with predictions
predictions_df.to_csv('predictions.csv', index=False)